# Introduction to Deep Learning & Neural Networks with Keras
Part B: Normalize the data

## Assignment Topic:
In this project, we will build a regression model using the Keras library to model the data about concrete compressive strength.

## Concrete Data:

The data can be found here: [Dataset Link](https://cocl.us/concrete_data)

The predictors in the data of concrete strength include:
1.   Cement
2.   Blast Furnace Slag
3.   Fly Ash
4.   Superplasticizer
5.   Coarse Aggregate
6.   Fine Aggregate

Download the concrete data file from the specified URL using wget

In [ ]:
# Uncomment the following line if you need to download the dataset directly from the download link
!wget "https://cocl.us/concrete_data" -O concrete_data.csv

--2023-10-03 17:21:51--  https://cocl.us/concrete_data
Resolving cocl.us (cocl.us)... 23.61.244.208, 23.35.110.231, 2600:140b:4800::173d:f439, ...
Connecting to cocl.us (cocl.us)|23.61.244.208|:443... connected.
HTTP request sent, awaiting response... 302 Moved Temporarily
Location: https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv [following]
--2023-10-03 17:21:52--  https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv
Resolving s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)... 67.228.254.196
Connecting to s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)|67.228.254.196|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 58988 (58K) [text/csv]
Saving to: ‘concrete_data.csv’

concrete_data.csv   100%[===================>]  57.61K   191KB/s    in 0.3s    

2023

Import necessary libraries and modules for data manipulation, splitting, metrics, and building a neural network model

In [ ]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense

Read the data from a CSV file named 'concrete_data' into a DataFrame (df)

In [ ]:
df = pd.read_csv('concrete_data.csv')

Display the first few rows of the DataFrame to get an initial glimpse of the data

In [ ]:
df.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30


Check the shape of the DataFrame to determine the number of rows and columns

In [ ]:
df.shape

(1030, 9)

Display summary statistics of the DataFrame

In [ ]:
df.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

Display information about the DataFrame, including data types and non-null counts

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Cement              1030 non-null   float64
 1   Blast Furnace Slag  1030 non-null   float64
 2   Fly Ash             1030 non-null   float64
 3   Water               1030 non-null   float64
 4   Superplasticizer    1030 non-null   float64
 5   Coarse Aggregate    1030 non-null   float64
 6   Fine Aggregate      1030 non-null   float64
 7   Age                 1030 non-null   int64  
 8   Strength            1030 non-null   float64
dtypes: float64(8), int64(1)
memory usage: 72.5 KB


Calculate the number of missing values (NaN) for each column in the DataFrame

In [ ]:
df.isna().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

Create a list of features by selecting all columns except 'Strength'

In [ ]:
features = [col for col in df.columns if col != 'Strength']
print('features:', features)

features: ['Cement', 'Blast Furnace Slag', 'Fly Ash', 'Water', 'Superplasticizer', 'Coarse Aggregate', 'Fine Aggregate', 'Age']


Extract the 'Strength' column as the target variable and convert it to a NumPy array

In [ ]:
target = df.Strength.to_numpy()

## Task for Part B
Normalize the features

In [ ]:
# Import the StandardScaler class from scikit-learn's preprocessing module
from sklearn.preprocessing import StandardScaler

# Initialize the StandardScaler
scaler = StandardScaler()

# Normalize the selected columns using fit_transform
df[features] = scaler.fit_transform(df[features])

Display the first few rows of the DataFrame to get a glimpse of the data after normalization

In [ ]:
df.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.477915           -0.856888 -0.847144 -0.916764         -0.620448   
1  2.477915           -0.856888 -0.847144 -0.916764         -0.620448   
2  0.491425            0.795526 -0.847144  2.175461         -1.039143   
3  0.491425            0.795526 -0.847144  2.175461         -1.039143   
4 -0.790459            0.678408 -0.847144  0.488793         -1.039143   

   Coarse Aggregate  Fine Aggregate       Age  Strength  
0          0.863154       -1.217670 -0.279733     79.99  
1          1.056164       -1.217670 -0.279733     61.89  
2         -0.526517       -2.240917  3.553066     40.27  
3         -0.526517       -2.240917  5.057677     41.05  
4          0.070527        0.647884  4.978487     44.30

## Task # 01
Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_split helper function from Scikit-learn.

In [ ]:
# Split the dataset into training and testing sets with a 70-30 split ratio and a fixed random seed
x_train, x_test, y_train, y_test = train_test_split(df[features], target, test_size=0.3, random_state=0)

print(f'number of rows and number of features in xtrain: {x_train.shape}')
print(f'number of rows in ytrain: {y_train.shape}')
print(f'number of rows and number of features in x_test: {x_test.shape}')
print(f'number of rows in y_test: {y_test.shape}')

number of rows and number of features in xtrain: (721, 8)
number of rows in ytrain: (721,)
number of rows and number of features in x_test: (309, 8)
number of rows in y_test: (309,)


In [ ]:
# Determine the number of columns in the feature dataset
n_cols = df[features].shape[1]
n_cols

8

## Task # 02
Train the model on the training data using 50 epochs.

### Create Regression Model

1.   One hidden layer of 10 nodes, and a ReLU activation function.
2.   Use the adam optimizer and the mean squared error as the loss function.



In [ ]:
# Define a function to create a regression model
def RegressionModel():
    # Create a sequential model
    model = Sequential()
    # One hidden layer of 10 nodes, and a ReLU activation function
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    # Add the output layer with 1 node
    model.add(Dense(1))

    # Compile the model using the Adam optimizer and mean squared error as the loss function
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [ ]:
# Create the regression model
model = RegressionModel()

In [ ]:
# Set the number of training epochs
n_epochs = 50

# Train the model on the training data
model.fit(x_train, y_train, epochs=n_epochs, verbose=1)

Epoch 1/50
23/23 [==============================] - 2s 7ms/step - loss: 1560.4819
Epoch 2/50
23/23 [==============================] - 0s 5ms/step - loss: 1544.2451
Epoch 3/50
23/23 [==============================] - 0s 5ms/step - loss: 1528.3724
Epoch 4/50
23/23 [==============================] - 0s 4ms/step - loss: 1511.8578
Epoch 5/50
23/23 [==============================] - 0s 6ms/step - loss: 1495.3135
Epoch 6/50
23/23 [==============================] - 0s 6ms/step - loss: 1478.2091
Epoch 7/50
23/23 [==============================] - 0s 5ms/step - loss: 1460.6077
Epoch 8/50
23/23 [==============================] - 0s 7ms/step - loss: 1442.1062
Epoch 9/50
23/23 [==============================] - 0s 4ms/step - loss: 1422.8590
Epoch 10/50
23/23 [==============================] - 0s 5ms/step - loss: 1402.3599
Epoch 11/50
23/23 [==============================] - 0s 5ms/step - loss: 1381.0901
Epoch 12/50
23/23 [==============================] - 0s 5ms/step - loss: 1358.5646
Epoch 13/50
2

## Task # 03
Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

In [ ]:
# Make predictions on the test data
y_hat = model.predict(x_test, verbose=0)

# Calculate and print the mean squared error between the predicted and actual target values
mse = mean_squared_error(y_test, y_hat)

print(f'Mean Squared Error: {mse}')

Mean Squared Error: 303.713646637444


## Task # 04
Repeat steps 1-3, 50 times, i.e., create a list of 50 mean squared errors.

In [ ]:
# Create an empty list to store mean squared errors
mse_list = []

# Set the number of runs
n_runs = 50

# Repeat the following steps 50 times, varying the random_state for data splitting
for i in range(n_runs):
    # Split the data into training and test sets (70% train, 30% test) with different random_state values
    x_train, x_test, y_train, y_test = train_test_split(df[features], target, test_size=0.3, random_state=i)

    # Create the regression model
    model = RegressionModel()

    # Train the model on the training data using n_epochs=50 epochs
    model.fit(x_train, y_train, epochs=n_epochs, verbose=0)

    # Make predictions on the test data
    y_hat = model.predict(x_test, verbose=0)

    # Evaluate the model on the test data and calculate mean squared error
    mean_square_error = mean_squared_error(y_test, y_hat)
    print(f'Mean Squared Error in Run {i + 1}: {mean_square_error}')

    # Append the calculated mean squared error to the list
    mse_list.append(mean_square_error)

Mean Squared Error in Run 1: 388.32956684762087
Mean Squared Error in Run 2: 495.89221756794495
Mean Squared Error in Run 3: 376.7229842646287
Mean Squared Error in Run 4: 287.06270678129164
Mean Squared Error in Run 5: 337.759216925084
Mean Squared Error in Run 6: 314.2480460577951
Mean Squared Error in Run 7: 261.98856773410347
Mean Squared Error in Run 8: 266.69079152006555
Mean Squared Error in Run 9: 514.5597326743039
Mean Squared Error in Run 10: 274.1268276363295
Mean Squared Error in Run 11: 720.4775912589786
Mean Squared Error in Run 12: 230.56362460110066
Mean Squared Error in Run 13: 273.5223566049686
Mean Squared Error in Run 14: 410.9584120517779
Mean Squared Error in Run 15: 407.79351611754066
Mean Squared Error in Run 16: 298.1497537133196
Mean Squared Error in Run 17: 291.773168709456
Mean Squared Error in Run 18: 487.39354924882593
Mean Squared Error in Run 19: 243.30229976652325
Mean Squared Error in Run 20: 347.41396849767796
Mean Squared Error in Run 21: 511.1158944

## Task # 05
Report the mean and the standard deviation of the mean squared errors.

In [ ]:
# Convert the list of mean squared errors into a NumPy array
mse_list = np.array(mse_list)

# Calculate the mean and standard deviation of the mean squared errors from all runs
mean = np.mean(mse_list)
standard_deviation = np.std(mse_list)

# Report by printing the mean and standard deviation of the mean squared errors
print(f'Mean of {n_runs} runs: {mean}')
print(f'Standard Deviation of {n_runs} runs: {standard_deviation}')

Mean of 50 runs: 334.0842202010648
Standard Deviation of 50 runs: 91.86404487875411


## How does the mean of the mean sqaured errors compare to that from Step A?
**Mean of MSE**: The second approach (with standardization) has a lower mean MSE (334.084) compared to the first approach (without standardization, 353.733). This suggests that, on average, the model with standardization provides slightly better predictive performance.

**Standard Deviation of MSE**: The standard deviation of MSE for the second approach is significantly lower (91.864) than that of the first approach (345.561). This means that the MSE values are more consistent and have less variability in the second approach, indicating that the model's performance is more stable across the 50 runs.

*Please take into consideration that when running these notebooks, the results may display variability owing to several factors. These factors may be random initialization, data partitioning, model training dynamics, variations in the versions of libraries such as Keras and TensorFlow, and/or additional external factors.*